src/Build/Evaluation/Expander.cs

```diff
@@ -1305,7 +1305,7 @@ private static object LookupProperty(IPropertyProvider<T> properties, string pro
 
                 object propertyValue;
 
-                if (property == null && MSBuildNameIgnoreCaseComparer.Equals("MSBuild", propertyName, startIndex, 7))
+                if (property == null && ((endIndex - startIndex) >= 7) && MSBuildNameIgnoreCaseComparer.Default.Equals("MSBuild", propertyName, startIndex, 7))
```

Comment:
This was breaking the assertion that the length passed in needs to fall within the range of "y", added to avoid reading past 'y'. This was just getting lucky because 'propertyName' here is guaranteed to be in the format of "$(Bar)" so we always ended up breaking on ")" and avoiding the AV.